[NEMO-79] Clean up the legacy Task

johnyangk opened the pull request:
JIRA: [NEMO-79: Clean up the legacy Task](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-79)

**Major changes:**
- Removes the 'old' Task interface and its implementations to make the 'new' Task (formerly TaskGroup) become the only Task in the codebase
- Specifically, changes the DAG in the 'new' Task from DAG<Task, RuntimeEdge<Task>> to DAG<IRVertex, RuntimeEdge<IRVertex>>
- Adds SourceVertex#clearInternalStates to clear the huge list of input splits held by BeamBoundedSourceVertex before sending the vertex to remote executors. Between clearing states of an existing vertex and creating a new vertex, I've chosen the former approach to ensure consistent use of the same IRVertex object across the compiler, the master, and the executors.

**Minor changes to note:**
- Make BeamBoundedSourceVertex remember sourceDescription for visualization purposes
- Changes related method and variable names appropriately
- Changes comments appropriately

**Tests for the changes:**
- N/A (no new feature was added)

**Other comments:**
- N/A

resolves [NEMO-79](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-79)

johnyangk commented:
@jeongyooneo Sorry, let me resolve the conflicts first. I'll let you know when the branch is ready to be reviewed.

johnyangk commented:
@jeongyooneo The branch is now ready for your review. 😄

seojangho commented:
Hi @johnyangk, thanks for the work!
It seems that changes in this PR makes the visualzer unable to display source information in physical plan. Hope it gets fixed before merging this PR. Thanks!

johnyangk commented:
Thanks @seojangho! Yes, I'll see if I can come up with something better than `SourceVertex#clearInternalStates` to avoid attaching all of the input splits to each task sent to an executor.

wonook commented on bin/json2dot.py:
```diff
@@ -64,8 +64,8 @@ def __init__(self, data):
         self.tasks = {}
-        for task in data['tasks']:
-            self.tasks[task['id']] = TaskState(task)
+        for irVertex in data['tasks']:
```
I think this part needs @seojangho 's review.

wonook commented on runtime/common/src/main/java/edu/snu/nemo/runtime/common/RuntimeIdGenerator.java:
```diff
@@ -76,36 +79,25 @@ public static String generateStageId(final Integer stageId) {
-  public static String generatePhysicalTaskId(final int index,
-                                              final String logicalTaskId) {
-    return logicalTaskId + PHYSICAL_TASK_ID_SPLITTER + index;
+  public static String generatePhysicalIRVertexId(final String irVertexId,
```
As far as I understand, this PR gets rid of the Tasks in runtime, and replaces it with the original IR vertices, which previous tasks have been referring to. However, this method creates ID for the Physical IR vertex which I'm unsure of its purpose. I think there should be rather a _Physical Vertex_ or something else and make it different from the original IR vertex, if it is not going to be used as it is used in the IR. Please check this part throughout the PR.

johnyangk replied:
Thanks! I've missed those. I've removed the related methods and variables, as they don't have a functional purpose.

wonook commented on runtime/common/src/main/java/edu/snu/nemo/runtime/common/RuntimeIdGenerator.java:
```diff
@@ -76,36 +79,25 @@ public static String generateStageId(final Integer stageId) {
    */
-  public static String generateTaskId(final int index,
-                                           final String stageId) {
+  public static String generateTaskId(final int index, final String stageId) {
```
This method makes it look like the tasks still exist. Please add a comment on the PR to better illustrate what this PR does, as it doesn't seem like it's getting rid of the tasks after all.

johnyangk replied:
This one is for the 'new' Task (formerly TaskGroup). I've updated the PR message to make it clear that the changes only remove the 'old' Task.

johnyangk commented:
@seojangho @wonook I've addressed your comments.

The issue with the visualizer has been fixed. This visualization of MapReduceITCase shows the source information correctly. 😄 

http://cmscluster.snu.ac.kr:50080/nemo-dag-out/680d8c49c7274afc99b5cf4071205aae.svg

jeongyooneo commented on common/src/main/java/edu/snu/nemo/common/ir/vertex/SourceVertex.java:
```diff
@@ -34,4 +34,9 @@
   public abstract List<Readable<O>> getReadables(int desiredNumOfSplits) throws Exception;
+
+  /**
+   * Clears internal states, must be called after getReadables().
```
Could you add the following description in this comment too?
Since the meaning and context of 'internal states' of `SourceVertex` isn't clear by itself.

> Adds SourceVertex#clearInternalStates to clear the huge list of input splits held by BeamBoundedSourceVertex before sending the vertex to remote executors.

johnyangk commented:
@jeongyooneo Thanks. I've addressed your comments.
